### ISSUE TITLE
ZeroMQ / JeroMQ - setReqCorrelate not working as expected

### ISSUE THREAD

**Issue by Nik-Novak** [zeromq/jeromq#772]
I recently discovered ZMQ and love it, except for the fact that I ran into this issue. Being that I'm relatively new, it's possible that I'm misunderstanding, but as far as I can see this seems to be a bug.

What I'm trying to accomplish:
I have a client and a server. The server (ZMQ.ROUTER) binds itself to an address and the client (ZMQ.REQ) connects to that address. The client is then able to send a message to the server, with a timeout duration for a response. It returns null if no response was received before timeout. For example's sake, the server just echoes back whatever the client sends, meaning that all frames remain the EXACT SAME as when they came in.

The main class creates a server (with a faulty boolean flag to simulate faulty server), and a client. The client sends a request to the server, waits for a response or maximum 5 seconds, then exits if a proper response was received. If a response was not received, the client retries the request in 2 seconds.

If I run this as-is, it works, but I cannot resend another request incase of failure (since REQ sockets must receive a reply before sending another request). Now I could just destroy() the socket and recreate it on failure, but I want to avoid this as it is inefficient and a dirty fix.

I found `ZMQ.Socket.setReqCorrelate(true)` and `ZMQ.setReqRelaxed(true)`, which combined perform exactly what I need, without destroying sockets, and while ensuring that previously timed-out responses dont come in late and falsely act as responses to newer requests. `ZMQ.setReqRelaxed(true)` allows a socket to send another request before receiving a reply, whereas `ZMQ.Socket.setReqCorrelate(true)` prepends each message with a RequestID frame, that ensures that the socket drops all late replies that don't have a matching RequestID frame. 

This is perfect! Except that it doesn't work. As soon as I uncomment `line 26` in the Client, it now drops all messages regardless of RequestID. The server receives the message just fine, processes it and sends it back to the client, but the client doesn't receive anything.

What I've tried:
I've tried playing around with the frames on the server-side, adding more copies of the requestID frame all over the message, Not using ZMsg and instead using raw byte[] arrays and multiple socket.receives. Nothing seems to work.

Any help would be greatly appreciated, please let me know if I'm doing something wrong.

Below is a sample application that shows this issue. Comment and uncomment `line 26` in Client.java to enable or disable setReqCorrelate. Simulate a faulty server (and connection retries) in Main.java with `FAULTY_SERVER=true;`  

Server.java: https://pastebin.com/p2S14uMq

Client.java: https://pastebin.com/epABqmpW

Main.java: https://pastebin.com/uCJrmMu9

Maven dependency:
```
<dependency>
  <groupId>org.zeromq</groupId>
  <artifactId>jeromq</artifactId>
  <version>0.5.2</version>
</dependency>
```

The only thing I could find related to this was a similar issue back in 2017 that was apparently fixed then undone:

FWIW, I think that change has been undone -- see https://github.com/zeromq/jeromq/issues/446 / https://github.com/fredoboulo/jeromq/commit/a8e715ee35a63fb8c9e356cf2628c5dedb669382#diff-ab533440049bba1d9014953971332f9aR410.

_Originally posted by @daveyarwood in https://github.com/zeromq/jeromq/issues/442#issuecomment-330218804_

**Comment by fredoboulo** [comment]
Hi, you're right, this PR was introduced later after the bump version and was never backported in Jeromq:
https://github.com/zeromq/libzmq/pull/1882

**Comment by trevorbernard** [comment]
@bluca What is the purpose of the following line? https://github.com/zeromq/libzmq/pull/1882/files#diff-80edcca992f1dac36f48cb59ffba75ccR286

**Comment by bluca** [comment]
Don't know, sorry - from the comment it looks like an optimization

**Comment by somdoron** [comment]
@trevorbernard I'm assuming because request-id is an integer, the line validates that the message part is in the correct size.

**Comment by somdoron** [comment]
It is actually makes sense, if correlate is on and the message-part is not in the correct size of request id then disconnect the peer.

**Comment by trevorbernard** [comment]
@Nik-Novak The fix should be in the latest 0.5.3-SNAPSHOT. Can you verify this fixed your issue?
